modules/tensor_mechanics/src/materials/ComputeIsotropicLinearElasticPFFractureStress.C

```diff
@@ -0,0 +1,141 @@
+#include "ComputeIsotropicLinearElasticPFFractureStress.h"
+
+template <>
+InputParameters
+validParams<ComputeIsotropicLinearElasticPFFractureStress>()
+{
+  InputParameters params = validParams<ComputeStressBase>();
+  params.addClassDescription("Phase-field fracture model energy contribution to fracture for "
+                             "isotropic elasticity and undamaged stress under compressive strain");
+  params.addRequiredCoupledVar("c", "Order parameter for damage");
+  params.addParam<Real>("kdamage", 1e-6, "Stiffness of damaged matrix");
+  params.addParam<MaterialPropertyName>(
+      "F_name", "E_el", "Name of material property storing the elastic energy");
+  return params;
+}
+
+ComputeIsotropicLinearElasticPFFractureStress::ComputeIsotropicLinearElasticPFFractureStress(
+    const InputParameters & parameters)
+  : ComputeStressBase(parameters),
+    _c(coupledValue("c")),
+    _kdamage(getParam<Real>("kdamage")),
+    _l(getMaterialProperty<Real>("l")),
+    _gc(getMaterialProperty<Real>("gc_prop")),
+    _F(declareProperty<Real>(getParam<MaterialPropertyName>("F_name"))),
+    _dFdc(declarePropertyDerivative<Real>(getParam<MaterialPropertyName>("F_name"),
+                                          getVar("c", 0)->name())),
+    _d2Fdc2(declarePropertyDerivative<Real>(
+        getParam<MaterialPropertyName>("F_name"), getVar("c", 0)->name(), getVar("c", 0)->name())),
+    _d2Fdcdstrain(declareProperty<RankTwoTensor>("d2Fdcdstrain")),
+    _dstress_dc(
+        declarePropertyDerivative<RankTwoTensor>(_base_name + "stress", getVar("c", 0)->name())),
+    _hist(declareProperty<Real>("hist")),
+    _hist_old(getMaterialPropertyOld<Real>("hist"))
+{
+}
+
+void
+ComputeIsotropicLinearElasticPFFractureStress::initQpStatefulProperties()
+{
+  _hist[_qp] = 0.0;
+}
+
+void
+ComputeIsotropicLinearElasticPFFractureStress::computeQpStress()
+{
+  const Real c = _c[_qp];
+
+  // Compute cfactor to have c > 1 behave the same as c = 1
+  Real cfactor = 0.0;
+  if (c < 1.0)
+    cfactor = 1.0;
+
+  // Isotropic elasticity is assumed and should be enforced
+  Real lambda = _elasticity_tensor[_qp](0, 0, 1, 1);
+  Real mu = _elasticity_tensor[_qp](0, 1, 0, 1);
+
+  // Compute eigenvectors and eigenvalues of mechanical strain
+  RankTwoTensor eigvec;
+  std::vector<Real> eigval(LIBMESH_DIM);
+  _mechanical_strain[_qp].symmetricEigenvaluesEigenvectors(eigval, eigvec);
+
+  // Calculate tensors of outerproduct of eigen vectors
+  std::vector<RankTwoTensor> etens(LIBMESH_DIM);
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+      for (unsigned int k = 0; k < LIBMESH_DIM; ++k)
+        etens[i](j, k) = eigvec(j, i) * eigvec(k, i);
+
+  // Separate out positive and negative eigen values
+  std::vector<Real> epos(LIBMESH_DIM), eneg(LIBMESH_DIM);
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+  {
+    epos[i] = (std::abs(eigval[i]) + eigval[i]) / 2.0;
```

Comment:
We have `MathUtils::positivePart(x)` and `MathUtils::negativePart(x)`, which we use in the other PFF kernels. This might improve readability.

Reply:
MathUtils is in the phase_field module. This material doesn't have access because it is in the tensor_mechanics module